File: Heap_sort.py
def heapify(sequence, n, i):
    largest = i
    l = 2 * i + 1
    r = 2 * i + 2

    if l < n and sequence[i] < sequence[l]:
        largest = l
    if r < n and sequence[largest] < sequence[r]:
        largest = r
    if largest != i:
        sequence[i], sequence[largest] = sequence[largest], sequence[i]
        heapify(sequence, n, largest)

def heapSort(sequence):
    n = len(sequence)
    for i in range(n // 2 - 1, -1, -1):
        heapify(sequence, n, i)

    for i in range(n - 1, 0, -1):
        sequence[i], sequence[0] = sequence[0], sequence[i]
        heapify(sequence, i, 0)

sequence = [12, 11, 13, 5, 6, 7]
heapSort(sequence)
n = len(sequence)
print("Sorted array is")
for i in range(n):
    print("%d" % sequence[i]),

File: HeapSort.java

public class HeapSort {

	public static void main(String[] args) {
		int[] inarray = {2,5,1,6,3,11,56,92,24,19,38,52,78,84,27,96,65};
		int n = inarray.length;
		for(int i = n/2 -1 ; i>=0 ; i--){
			heapify(inarray, n, i);
		}
		for(int i : inarray){
			System.out.println(i);
		}
		System.out.println("---------------");
		for(int i = n -1 ; i>=0 ; i--){
			
			int temp = inarray[0];
			inarray[0] = inarray[i];
			inarray[i] = temp;
			heapify(inarray, i, 0);
		}
		
		for(int i : inarray){
			System.out.println(i);
		}
		
	}
	
	public static void heapify(int[] inarray, int n , int i){
		
		int largest = i;
		int l = 2*i+1;
		int r = 2*i+2;
		
		if(l<n && inarray[l]>inarray[largest]){
			largest = l;
		}
		
		if(r<n && inarray[r]>inarray[largest]){
			largest = r;
		}
		
		if(largest != i){
			int temp = inarray[i];
			inarray[i] = inarray[largest];
			inarray[largest] = temp;
			heapify(inarray, n, largest);
		}
	}
}
